nixos/modules/services/networking/quassel.nix

```diff
@@ -6,134 +6,498 @@ let
-        type = types.nullOr types.str;
-        default = null;
-        description = lib.mdDoc ''
-          Path to the certificate used for SSL connections with clients.
-        '';
-      };
-
-      requireSSL = mkOption {
-        type = types.bool;
-        default = false;
-        description = lib.mdDoc ''
-          Require SSL for connections from clients.
-        '';
-      };
+      package = lib.mkPackageOptionMD pkgs "quasselDaemon" { };
 
-      package = mkOption {
-        type = types.package;
-        default = pkgs.quasselDaemon;
-        defaultText = literalExpression "pkgs.quasselDaemon";
+      user = mkOption {
+        type = types.str;
+        default = "quassel";
         description = lib.mdDoc ''
-          The package of the quassel daemon.
+          The user the Quassel daemon should run as. By default a systemd DynamicUser
+          is used with the name specified here. DynamicUser functionality will be
+          automatically disabled if the specified user already exists.
         '';
       };
 
-      interfaces = mkOption {
-        type = types.listOf types.str;
-        default = [ "127.0.0.1" ];
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
         description = lib.mdDoc ''
-          The interfaces the Quassel daemon will be listening to.  If `[ 127.0.0.1 ]`,
-          only clients on the local host can connect to it; if `[ 0.0.0.0 ]`, clients
-          can access it from any network interface.
-        '';
-      };
+          Path to an environment file loaded for the quassel service.
 
-      portNumber = mkOption {
-        type = types.port;
-        default = 4242;
-        description = lib.mdDoc ''
-          The port number the Quassel daemon will be listening to.
-        '';
-      };
+          This can be used to securely store tokens and secrets outside of the world-readable Nix store.
+          For example to inject the `DB_PSQL_PASSWORD` and `AUTH_LDAP_BIND_PASSWORD` variables instead
+          of setting {option}`services.quassel.settings.db.psql.password` and {option}`services.quassel.settings.auth.ldap.bindPassword`.
 
-      dataDir = mkOption {
-        default = "/home/${user}/.config/quassel-irc.org";
-        defaultText = literalExpression ''
-          "/home/''${config.${opt.user}}/.config/quassel-irc.org"
+          ::: {.note}
+          Since this file is read by systemd, it may have permission 0400 and be owned by root.
+          :::
         '';
-        type = types.str;
-        description = lib.mdDoc ''
-          The directory holding configuration files, the SQlite database and the SSL Cert.
+        example = ''
+          # Example Content of the File
+          DB_PSQL_PASSWORD=changeme
+          AUTH_LDAP_BIND_PASSWORD=changemetoo
         '';
       };
 
-      user = mkOption {
-        default = null;
-        type = types.nullOr types.str;
+      settings = mkOption {
```

Comment:
Considering this isn't a freeform module, I think it's a bit overkill to use submodules for every single layer of nesting. It would be a lot cleaner and easier to deal with, if this was just a big attrset of options.

Could you rewrite it to work like this?

```nix
      settings = {
        listen = mkOption {
          type = types.listOf types.str;
          default = [ "127.0.0.1" "::1" ];
          description = lib.mdDoc ''
            The address(es) quasselcore will listen on.
          '';
        };
        ...
        ident = {
          enable = lib.mkEnableOption (lib.mdDoc "internal ident daemon.");
          ...
        };
        ...
      };
```